### PR TITLE
HHH-16423 + HHH-16447 + HHH-16453 + HHH-16496

### PR DESCRIPTION
**Pull request by dreab8**
- https://hibernate.atlassian.net/browse/HHH-16423 LazyInitializationException when calling em.refresh() on entity with lazy and eager children
- https://hibernate.atlassian.net/browse/HHH-16447 Lazy load of collection fails after EntityManager.refresh
- https://hibernate.atlassian.net/browse/HHH-16453 Loading an Entity with two eager collections, produces duplicates when one of the collection is a a bag
- https://hibernate.atlassian.net/browse/HHH-16496 Join fetching nested @OneToMany collections causes result multiplication

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/sql/results/graph/entity/AbstractEntityInitializer.java [hunk 1]
```diff
@@ -530,7 +531,20 @@ else if ( entityInstanceFromExecutionContext != null ) {
 		}
 	}
 
+	protected void registerLoadingEntityInstanceFromExecutionContext(
+			RowProcessingState rowProcessingState,
+			Object instance) {
+	}
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/sql/results/graph/entity/AbstractEntityInitializer.java:537]: Can you please add a comment here that explains why we don't want to register the entity instance?

File: hibernate-core/src/main/java/org/hibernate/sql/results/graph/entity/AbstractEntityInitializer.java [hunk 2]
```diff
@@ -530,7 +531,20 @@ else if ( entityInstanceFromExecutionContext != null ) {
...
+
 	protected Object getEntityInstanceFromExecutionContext(RowProcessingState rowProcessingState) {
+		final ExecutionContext executionContext = rowProcessingState.getJdbcValuesSourceProcessingState()
+				.getExecutionContext();
+		final Object entityInstanceFromExecutionContext = executionContext.getEntityInstance();
+		if ( entityInstanceFromExecutionContext != null
+				&& getEntityKey().getIdentifier().equals( executionContext.getEntityId() )
+				&& getConcreteDescriptor().getJavaType().getJavaType().equals( entityInstanceFromExecutionContext.getClass() ) ) {
```
[the remark this answers is not in the document]
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/sql/results/graph/entity/AbstractEntityInitializer.java:545]: not sure we always have the `EntityPersister` when we create the `ExecutionContext`

**Review by beikov — COMMENTED** [review]
I looked though the PR but the changes in `LoaderSelectBuilder` don't look quite right to me. 

Isn't fetching a bag always a problem if we already fetched any sort of collection? IMO we should rather track the possible cardinality states with an enum field `ResultCardinality` e.g. `SINGLE`, `DUPLICATE` and `BAG`. When join fetching a collection we transition to `DUPLICATE` or `BAG` respectively. A bag can only be join fetched if the state is `SINGLE`. A non-bag collection can always be join fetched, except the state is `BAG`. This way, it also makes sense to try join fetching bags last.

**Comment by dreab8** [comment]
tha aim of this change is to not join any bag if there are other joined non-bag collections, nut what i want to avoid is to join a bag and then not to the following join N non-bag collections but instead join the N non-bag collection and not join the bag

**Pull request merged**